
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Working with PDB files &#8212; Introduction to Bio Libraries</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=62ba249389abaaa9ffc34bf36a076bdc1d65ee18" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=f31d14ad54b65d19161ba51d4ffff3a77ae00456"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Phylogenetic analysis" href="Phylo.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo_final_horizontal.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Introduction to Bio Libraries</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="BioPython_Intro.html">
   Introduction to BioPython
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Motif_analysis.html">
   Motif analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="Phylo.html">
   Phylogenetic analysis
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Working with PDB files
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
             <style> .text1{font-size=18pt;} .footer1{ margin: 2em; text-align: center; } .social_media{ color: #9c51e0; font-size: 1.5em; word-spacing: 0.5em; } </style> <script src="https://kit.fontawesome.com/578801937d.js" crossorigin="anonymous"></script> <link rel="preconnect" href="https://fonts.googleapis.com"> <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin> <link href="https://fonts.googleapis.com/css2?family=Geo&family=Lato&display=swap" rel="stylesheet"> <div class="footer1"> <span class="text1">Connect with </span> <span class="text1" style="color:#9c51e0;font-family:Geo;font-size:14pt">bioinfo.guru</span> <br> <span class="social_media"> <a href="https://www.facebook.com/profile.php?id=100078250833738"><i class="fa-brands fa-facebook-square social_media"></i></a> <a href="https://www.instagram.com/bioinfo.guru/"><i class="fa-brands fa-instagram-square social_media"></i></a> <a href="https://www.linkedin.com/in/bioinfo-guru-68a8b7231/"><i class="fa-brands fa-linkedin social_media"></i></a> <a href="https://twitter.com/bioinfo_guru"><i class="fa-brands fa-twitter-square social_media"></i></a> </span> </div> 
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Working with PDB files
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pdb-format">
     PDB format
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#structure-object">
     Structure object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#shortcuts-for-accessing-structure-elements">
     Shortcuts for accessing structure elements
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#nested-lists">
       Nested lists
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#residue-object">
     Residue object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#atom-object">
     Atom object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#distance-calculations">
     Distance Calculations
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#binana">
   Binana
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Working with PDB files</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Working with PDB files
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#pdb-format">
     PDB format
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#structure-object">
     Structure object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#shortcuts-for-accessing-structure-elements">
     Shortcuts for accessing structure elements
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#nested-lists">
       Nested lists
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#residue-object">
     Residue object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#atom-object">
     Atom object
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#distance-calculations">
     Distance Calculations
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#binana">
   Binana
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <div class="section" id="working-with-pdb-files">
<h1>Working with PDB files<a class="headerlink" href="#working-with-pdb-files" title="Permalink to this headline">#</a></h1>
<div class="section" id="pdb-format">
<h2>PDB format<a class="headerlink" href="#pdb-format" title="Permalink to this headline">#</a></h2>
<p>Biomolecular structures can be saved in different formats such as PDB, mmCIF, XML, etc. The PDB format is the most common format and is supported by all visualization and docking programs. The PDB class of Bio library has a <code class="docutils literal notranslate"><span class="pre">PDBParser()</span></code> function to instantiate a PDB parser object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># %load 7jtl_atoms.pdb</span>
<span class="n">HEADER</span>    <span class="n">VIRAL</span> <span class="n">PROTEIN</span>                           <span class="mi">18</span><span class="o">-</span><span class="n">AUG</span><span class="o">-</span><span class="mi">20</span>   <span class="mi">7</span><span class="n">JTL</span>              
<span class="n">TITLE</span>     <span class="n">STRUCTURE</span> <span class="n">OF</span> <span class="n">SARS</span><span class="o">-</span><span class="n">COV</span><span class="o">-</span><span class="mi">2</span> <span class="n">ORF8</span> <span class="n">ACCESSORY</span> <span class="n">PROTEIN</span>                        
<span class="n">COMPND</span>    <span class="n">MOL_ID</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span>                                                            
<span class="n">COMPND</span>   <span class="mi">2</span> <span class="n">MOLECULE</span><span class="p">:</span> <span class="n">ORF8</span> <span class="n">PROTEIN</span><span class="p">;</span>                                              
<span class="n">COMPND</span>   <span class="mi">3</span> <span class="n">CHAIN</span><span class="p">:</span> <span class="n">A</span><span class="p">,</span> <span class="n">B</span><span class="p">;</span>                                                         
<span class="n">COMPND</span>   <span class="mi">4</span> <span class="n">SYNONYM</span><span class="p">:</span> <span class="n">ORF8</span><span class="p">,</span> <span class="n">NON</span><span class="o">-</span><span class="n">STRUCTURAL</span> <span class="n">PROTEIN</span> <span class="mi">8</span><span class="p">,</span> <span class="n">NS8</span><span class="p">;</span>                        
<span class="n">COMPND</span>   <span class="mi">5</span> <span class="n">ENGINEERED</span><span class="p">:</span> <span class="n">YES</span>                                                      
<span class="n">SOURCE</span>    <span class="n">MOL_ID</span><span class="p">:</span> <span class="mi">1</span><span class="p">;</span>                                                            
<span class="n">SOURCE</span>   <span class="mi">2</span> <span class="n">ORGANISM_SCIENTIFIC</span><span class="p">:</span> <span class="n">SEVERE</span> <span class="n">ACUTE</span> <span class="n">RESPIRATORY</span> <span class="n">SYNDROME</span> <span class="n">CORONAVIRUS</span>   
<span class="n">SOURCE</span>   <span class="mi">3</span> <span class="mi">2</span><span class="p">;</span>                                                                   
<span class="n">SOURCE</span>   <span class="mi">4</span> <span class="n">ORGANISM_COMMON</span><span class="p">:</span> <span class="mi">2019</span><span class="o">-</span><span class="n">NCOV</span><span class="p">;</span>                                          
<span class="n">SOURCE</span>   <span class="mi">5</span> <span class="n">ORGANISM_TAXID</span><span class="p">:</span> <span class="mi">2697049</span><span class="p">;</span>                                             
<span class="n">SOURCE</span>   <span class="mi">6</span> <span class="n">EXPRESSION_SYSTEM</span><span class="p">:</span> <span class="n">ESCHERICHIA</span> <span class="n">COLI</span><span class="p">;</span>                                 
<span class="n">SOURCE</span>   <span class="mi">7</span> <span class="n">EXPRESSION_SYSTEM_TAXID</span><span class="p">:</span> <span class="mi">562</span>                                         
<span class="n">ATOM</span>      <span class="mi">1</span>  <span class="n">N</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">18</span>      <span class="mf">40.917</span>  <span class="mf">33.173</span> <span class="mf">126.601</span>  <span class="mf">1.00</span> <span class="mf">73.92</span>           <span class="n">N</span>  
<span class="n">ATOM</span>      <span class="mi">2</span>  <span class="n">CA</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">18</span>      <span class="mf">42.025</span>  <span class="mf">33.162</span> <span class="mf">125.645</span>  <span class="mf">1.00</span> <span class="mf">73.89</span>           <span class="n">C</span>  
<span class="n">ATOM</span>      <span class="mi">3</span>  <span class="n">C</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">18</span>      <span class="mf">41.754</span>  <span class="mf">32.204</span> <span class="mf">124.493</span>  <span class="mf">1.00</span> <span class="mf">67.88</span>           <span class="n">C</span>  
<span class="n">ATOM</span>      <span class="mi">4</span>  <span class="n">O</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">18</span>      <span class="mf">42.414</span>  <span class="mf">31.177</span> <span class="mf">124.374</span>  <span class="mf">1.00</span> <span class="mf">66.48</span>           <span class="n">O</span>  
<span class="n">ATOM</span>      <span class="mi">5</span>  <span class="n">CB</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">18</span>      <span class="mf">43.339</span>  <span class="mf">32.782</span> <span class="mf">126.337</span>  <span class="mf">1.00</span> <span class="mf">73.82</span>           <span class="n">C</span>  
<span class="n">ATOM</span>      <span class="mi">6</span>  <span class="n">N</span>   <span class="n">GLU</span> <span class="n">A</span>  <span class="mi">19</span>      <span class="mf">40.764</span>  <span class="mf">32.527</span> <span class="mf">123.665</span>  <span class="mf">1.00</span> <span class="mf">60.90</span>           <span class="n">N</span>  
<span class="n">ATOM</span>      <span class="mi">7</span>  <span class="n">CA</span>  <span class="n">GLU</span> <span class="n">A</span>  <span class="mi">19</span>      <span class="mf">40.434</span>  <span class="mf">31.695</span> <span class="mf">122.517</span>  <span class="mf">1.00</span> <span class="mf">50.80</span>           <span class="n">C</span>  
<span class="n">ATOM</span>      <span class="mi">8</span>  <span class="n">C</span>   <span class="n">GLU</span> <span class="n">A</span>  <span class="mi">19</span>      <span class="mf">40.928</span>  <span class="mf">32.273</span> <span class="mf">121.201</span>  <span class="mf">1.00</span> <span class="mf">50.81</span>           <span class="n">C</span>  
<span class="n">ATOM</span>      <span class="mi">9</span>  <span class="n">O</span>   <span class="n">GLU</span> <span class="n">A</span>  <span class="mi">19</span>      <span class="mf">41.122</span>  <span class="mf">31.517</span> <span class="mf">120.240</span>  <span class="mf">1.00</span> <span class="mf">47.32</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">10</span>  <span class="n">CB</span>  <span class="n">GLU</span> <span class="n">A</span>  <span class="mi">19</span>      <span class="mf">38.918</span>  <span class="mf">31.467</span> <span class="mf">122.431</span>  <span class="mf">1.00</span> <span class="mf">55.27</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">11</span>  <span class="n">N</span>   <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">41.101</span>  <span class="mf">33.590</span> <span class="mf">121.115</span>  <span class="mf">1.00</span> <span class="mf">51.03</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">12</span>  <span class="n">CA</span>  <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">41.551</span>  <span class="mf">34.244</span> <span class="mf">119.893</span>  <span class="mf">1.00</span> <span class="mf">41.77</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">13</span>  <span class="n">C</span>   <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">42.699</span>  <span class="mf">35.209</span> <span class="mf">120.189</span>  <span class="mf">1.00</span> <span class="mf">40.41</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">14</span>  <span class="n">O</span>   <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">42.709</span>  <span class="mf">35.887</span> <span class="mf">121.221</span>  <span class="mf">1.00</span> <span class="mf">46.45</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">15</span>  <span class="n">CB</span>  <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">40.393</span>  <span class="mf">34.986</span> <span class="mf">119.225</span>  <span class="mf">1.00</span> <span class="mf">46.48</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">16</span>  <span class="n">SG</span>  <span class="n">CYS</span> <span class="n">A</span>  <span class="mi">20</span>      <span class="mf">40.867</span>  <span class="mf">35.722</span> <span class="mf">117.656</span>  <span class="mf">1.00</span> <span class="mf">56.21</span>           <span class="n">S</span>  
<span class="n">ATOM</span>     <span class="mi">17</span>  <span class="n">N</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">43.669</span>  <span class="mf">35.254</span> <span class="mf">119.283</span>  <span class="mf">1.00</span> <span class="mf">38.94</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">18</span>  <span class="n">CA</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">44.851</span>  <span class="mf">36.104</span> <span class="mf">119.410</span>  <span class="mf">1.00</span> <span class="mf">35.69</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">19</span>  <span class="n">C</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">44.916</span>  <span class="mf">36.902</span> <span class="mf">118.118</span>  <span class="mf">1.00</span> <span class="mf">35.79</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">20</span>  <span class="n">O</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">45.172</span>  <span class="mf">36.333</span> <span class="mf">117.051</span>  <span class="mf">1.00</span> <span class="mf">32.57</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">21</span>  <span class="n">CB</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">46.118</span>  <span class="mf">35.269</span> <span class="mf">119.634</span>  <span class="mf">1.00</span> <span class="mf">33.65</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">22</span>  <span class="n">OG</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">21</span>      <span class="mf">47.328</span>  <span class="mf">36.023</span> <span class="mf">119.512</span>  <span class="mf">1.00</span> <span class="mf">33.95</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">23</span>  <span class="n">N</span>   <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">44.643</span>  <span class="mf">38.202</span> <span class="mf">118.205</span>  <span class="mf">1.00</span> <span class="mf">36.94</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">24</span>  <span class="n">CA</span>  <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">44.631</span>  <span class="mf">39.082</span> <span class="mf">117.041</span>  <span class="mf">1.00</span> <span class="mf">35.49</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">25</span>  <span class="n">C</span>   <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">45.993</span>  <span class="mf">39.763</span> <span class="mf">116.942</span>  <span class="mf">1.00</span> <span class="mf">34.55</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">26</span>  <span class="n">O</span>   <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">46.412</span>  <span class="mf">40.451</span> <span class="mf">117.875</span>  <span class="mf">1.00</span> <span class="mf">37.54</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">27</span>  <span class="n">CB</span>  <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">43.520</span>  <span class="mf">40.122</span> <span class="mf">117.155</span>  <span class="mf">1.00</span> <span class="mf">38.67</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">28</span>  <span class="n">CG</span>  <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">43.475</span>  <span class="mf">41.147</span> <span class="mf">116.024</span>  <span class="mf">1.00</span> <span class="mf">41.31</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">29</span>  <span class="n">CD1</span> <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">43.308</span>  <span class="mf">40.438</span> <span class="mf">114.693</span>  <span class="mf">1.00</span> <span class="mf">37.10</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">30</span>  <span class="n">CD2</span> <span class="n">LEU</span> <span class="n">A</span>  <span class="mi">22</span>      <span class="mf">42.386</span>  <span class="mf">42.162</span> <span class="mf">116.241</span>  <span class="mf">1.00</span> <span class="mf">46.10</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">31</span>  <span class="n">N</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">46.676</span>  <span class="mf">39.584</span> <span class="mf">115.820</span>  <span class="mf">1.00</span> <span class="mf">32.31</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">32</span>  <span class="n">CA</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">48.026</span>  <span class="mf">40.101</span> <span class="mf">115.679</span>  <span class="mf">1.00</span> <span class="mf">35.96</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">33</span>  <span class="n">C</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">48.129</span>  <span class="mf">40.869</span> <span class="mf">114.379</span>  <span class="mf">1.00</span> <span class="mf">36.92</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">34</span>  <span class="n">O</span>   <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">47.308</span>  <span class="mf">40.706</span> <span class="mf">113.477</span>  <span class="mf">1.00</span> <span class="mf">34.85</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">35</span>  <span class="n">CB</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">49.080</span>  <span class="mf">38.978</span> <span class="mf">115.752</span>  <span class="mf">1.00</span> <span class="mf">33.22</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">36</span>  <span class="n">CG</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">49.081</span>  <span class="mf">38.307</span> <span class="mf">117.122</span>  <span class="mf">1.00</span> <span class="mf">36.84</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">37</span>  <span class="n">CD</span>  <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">50.113</span>  <span class="mf">37.220</span> <span class="mf">117.258</span>  <span class="mf">1.00</span> <span class="mf">37.90</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">38</span>  <span class="n">OE1</span> <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">51.178</span>  <span class="mf">37.277</span> <span class="mf">116.645</span>  <span class="mf">1.00</span> <span class="mf">33.05</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">39</span>  <span class="n">NE2</span> <span class="n">GLN</span> <span class="n">A</span>  <span class="mi">23</span>      <span class="mf">49.808</span>  <span class="mf">36.219</span> <span class="mf">118.072</span>  <span class="mf">1.00</span> <span class="mf">35.01</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">40</span>  <span class="n">N</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">49.141</span>  <span class="mf">41.729</span> <span class="mf">114.310</span>  <span class="mf">1.00</span> <span class="mf">33.69</span>           <span class="n">N</span>  
<span class="n">ATOM</span>     <span class="mi">41</span>  <span class="n">CA</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">49.355</span>  <span class="mf">42.636</span> <span class="mf">113.187</span>  <span class="mf">1.00</span> <span class="mf">36.00</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">42</span>  <span class="n">C</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">50.733</span>  <span class="mf">42.408</span> <span class="mf">112.590</span>  <span class="mf">1.00</span> <span class="mf">38.21</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">43</span>  <span class="n">O</span>   <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">51.714</span>  <span class="mf">42.271</span> <span class="mf">113.320</span>  <span class="mf">1.00</span> <span class="mf">33.70</span>           <span class="n">O</span>  
<span class="n">ATOM</span>     <span class="mi">44</span>  <span class="n">CB</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">49.242</span>  <span class="mf">44.119</span> <span class="mf">113.611</span>  <span class="mf">1.00</span> <span class="mf">38.78</span>           <span class="n">C</span>  
<span class="n">ATOM</span>     <span class="mi">45</span>  <span class="n">OG</span>  <span class="n">SER</span> <span class="n">A</span>  <span class="mi">24</span>      <span class="mf">47.976</span>  <span class="mf">44.409</span> <span class="mf">114.168</span>  <span class="mf">1.00</span> <span class="mf">44.24</span>           <span class="n">O</span>  
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">PDBParser()</span></code> function in Bio.PDB instantiates an object that have functions to read structure files. The <code class="docutils literal notranslate"><span class="pre">get_structure()</span></code> function takes two arguments - an id and a file name (or an open filehandle).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.PDB</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">parser</span> <span class="o">=</span> <span class="n">PDBParser</span><span class="p">()</span>
<span class="n">structure</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">get_structure</span><span class="p">(</span><span class="s2">&quot;Protease&quot;</span><span class="p">,</span> <span class="s2">&quot;7jtl.pdb&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>c:\Users\ashish\anaconda3\envs\bio1\lib\site-packages\Bio\PDB\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 3855.
  warnings.warn(
c:\Users\ashish\anaconda3\envs\bio1\lib\site-packages\Bio\PDB\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain B is discontinuous at line 3963.
  warnings.warn(
</pre></div>
</div>
</div>
</div>
<p>The annotations for the structue are available in the <code class="docutils literal notranslate"><span class="pre">header</span></code> attribute of structure object. This retuns a dictionary of different keywords associated with the structure and their correponding values as shown below.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">header</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">header</span><span class="p">[</span><span class="s2">&quot;resolution&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">structure</span><span class="o">.</span><span class="n">header</span><span class="p">[</span><span class="s2">&quot;keywords&quot;</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>structure of sars-cov-2 orf8 accessory protein
2.04
sars-cov-2, sars2, covid-19, coronavirus, accessory protein, host- factor restriction, rna virus, immune evasion, mhc-i, open reading frame 8, viral protein
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="structure-object">
<h2>Structure object<a class="headerlink" href="#structure-object" title="Permalink to this headline">#</a></h2>
<p>The structure object has a hierarchical orgranization which can be thought of as a nested list (list of lists).</p>
<p>Structure <span class="math notranslate nohighlight">\(\longrightarrow\)</span> Model <span class="math notranslate nohighlight">\(\longrightarrow\)</span> Chain <span class="math notranslate nohighlight">\(\longrightarrow\)</span> Residue <span class="math notranslate nohighlight">\(\longrightarrow\)</span> Atom</p>
<p>Each object in this hierarchy has certain functions associated with it to get information about that object.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#for model in structure:</span>
<span class="c1">#    print(model)</span>
<span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">chain</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Chain id=A&gt;
&lt;Chain id=B&gt;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ctr</span><span class="o">=</span><span class="mi">0</span>
<span class="k">for</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">structure</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">chain</span> <span class="ow">in</span> <span class="n">model</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">chain</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">residue</span><span class="p">:</span>
                <span class="k">if</span><span class="p">(</span><span class="n">ctr</span><span class="o">==</span><span class="mi">5</span><span class="p">):</span>
                    <span class="k">break</span>
                <span class="nb">print</span><span class="p">(</span><span class="n">atom</span><span class="p">)</span>
                <span class="n">ctr</span><span class="o">+=</span><span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Atom N&gt;
&lt;Atom CA&gt;
&lt;Atom C&gt;
&lt;Atom O&gt;
&lt;Atom CB&gt;
</pre></div>
</div>
</div>
</div>
<p>Getting amino acid sequence from the pdb file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ctr</span><span class="o">=</span><span class="mi">0</span>
<span class="n">prot_seq</span><span class="o">=</span><span class="s2">&quot;&quot;</span>
<span class="k">for</span> <span class="n">residue</span> <span class="ow">in</span> <span class="n">model</span><span class="o">.</span><span class="n">get_residues</span><span class="p">():</span>
    <span class="k">if</span><span class="p">(</span><span class="n">ctr</span><span class="o">==</span><span class="mi">5</span><span class="p">):</span>
        <span class="k">break</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">residue</span><span class="o">.</span><span class="n">resname</span><span class="p">,</span> <span class="n">residue</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">prot_seq</span><span class="o">+=</span><span class="n">residue</span><span class="o">.</span><span class="n">resname</span>
    <span class="n">ctr</span><span class="o">+=</span><span class="mi">1</span>
<span class="nb">print</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GLN 18
GLU 19
CYS 20
SER 21
LEU 22
GLNGLUCYSSERLEU
</pre></div>
</div>
</div>
</div>
<p>To convert the sequence in three letter code to single letter code, we can use <code class="docutils literal notranslate"><span class="pre">SeqUtils</span></code> package available in biopython. The <code class="docutils literal notranslate"><span class="pre">seq1()</span></code> function returns single letter sequence for a three letter sequence. Similarly, to convert single letter sequence to three letter we have <code class="docutils literal notranslate"><span class="pre">seq3()</span></code> function.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">seq1</span>
<span class="n">seq1</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&#39;QECSL&#39;
</pre></div>
</div>
</div>
</div>
<p>In case the structure has some modified residues then custom mapping of three letter code to single letter code can be done using the <code class="docutils literal notranslate"><span class="pre">custom_map</span></code> keyword argument to <code class="docutils literal notranslate"><span class="pre">seq1()</span></code> function. E.g., let say a protein has a phosphorylated serine residue for which the three letter code is SEP and there is no standard single letter code for this. We can add custom mapping for SEP to S as follows.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">prot_seq</span><span class="o">+=</span><span class="s2">&quot;SEP&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">))</span> <span class="c1">#without custom map</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">(</span><span class="n">prot_seq</span><span class="p">,</span><span class="n">custom_map</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;SEP&quot;</span><span class="p">:</span> <span class="s2">&quot;S&quot;</span><span class="p">}))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>QECSLX
QECSLS
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="shortcuts-for-accessing-structure-elements">
<h2>Shortcuts for accessing structure elements<a class="headerlink" href="#shortcuts-for-accessing-structure-elements" title="Permalink to this headline">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Selection()</span></code> class in Bio.PDB has some useful functions to select specific data from a structure object. E.g. the <code class="docutils literal notranslate"><span class="pre">unfold_entities()</span></code> function provides a shortcut to unpack a structure object at the desired level within the structure hierarchy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atom_list</span> <span class="o">=</span> <span class="n">Selection</span><span class="o">.</span><span class="n">unfold_entities</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">all_coords</span> <span class="o">=</span> <span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">coord</span> <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">atom_list</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">all_coords</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[array([ 40.917,  33.173, 126.601], dtype=float32), array([ 42.025,  33.162, 125.645], dtype=float32), array([ 41.754,  32.204, 124.493], dtype=float32)]
</pre></div>
</div>
</div>
</div>
<div class="section" id="nested-lists">
<h3>Nested lists<a class="headerlink" href="#nested-lists" title="Permalink to this headline">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">list1</span> <span class="o">=</span> <span class="p">[[[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">]],[[</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">],[</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]]]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">list1</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atom</span> <span class="o">=</span> <span class="n">structure</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;A&quot;</span><span class="p">][</span><span class="mi">100</span><span class="p">][</span><span class="s2">&quot;CA&quot;</span><span class="p">]</span>
<span class="n">atom</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;Atom CA&gt;
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="residue-object">
<h2>Residue object<a class="headerlink" href="#residue-object" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res_list</span> <span class="o">=</span> <span class="n">Selection</span><span class="o">.</span><span class="n">unfold_entities</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="s2">&quot;R&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res_list</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_resname</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res_list</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_list</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">res_list</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>SER
[&lt;Atom N&gt;, &lt;Atom CA&gt;, &lt;Atom C&gt;, &lt;Atom O&gt;, &lt;Atom CB&gt;, &lt;Atom OG&gt;]
21
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="atom-object">
<h2>Atom object<a class="headerlink" href="#atom-object" title="Permalink to this headline">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atoms_resi3</span> <span class="o">=</span> <span class="n">res_list</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_list</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">atoms_resi3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">atoms_resi3</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_name</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="n">atoms_resi3</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">get_coord</span><span class="p">())</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&lt;Atom N&gt;, &lt;Atom CA&gt;, &lt;Atom C&gt;, &lt;Atom O&gt;, &lt;Atom CB&gt;, &lt;Atom OG&gt;]
CA
[ 44.851  36.104 119.41 ]
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="distance-calculations">
<h2>Distance Calculations<a class="headerlink" href="#distance-calculations" title="Permalink to this headline">#</a></h2>
<p>We can use the <code class="docutils literal notranslate"><span class="pre">-</span></code> operator with two atoms to get the distance between the atoms.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ca_3</span> <span class="o">=</span> <span class="n">res_list</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">get_list</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">ca_15</span> <span class="o">=</span> <span class="n">res_list</span><span class="p">[</span><span class="mi">15</span><span class="p">]</span><span class="o">.</span><span class="n">get_list</span><span class="p">()[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">distance</span> <span class="o">=</span> <span class="n">ca_3</span> <span class="o">-</span> <span class="n">ca_15</span>
<span class="nb">print</span><span class="p">(</span><span class="n">distance</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>10.335114
</pre></div>
</div>
</div>
</div>
<p>Calulating Ca-Ca distance matrix</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">atom_list</span> <span class="o">=</span> <span class="n">res_list</span> <span class="o">=</span> <span class="n">Selection</span><span class="o">.</span><span class="n">unfold_entities</span><span class="p">(</span><span class="n">structure</span><span class="p">,</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
<span class="n">ca_ca_dist</span> <span class="o">=</span> <span class="p">[]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number of atoms&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">atom_list</span><span class="p">))</span>
<span class="n">ca_atom_list</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">atom_list</span> <span class="k">if</span> <span class="s2">&quot;CA&quot;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of Ca atoms&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ca_atom_list</span><span class="p">))</span>

<span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">ca_atom_list</span><span class="p">)):</span>
    <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">x</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">ca_atom_list</span><span class="p">)):</span>
        <span class="n">ca_ca_dist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ca_atom_list</span><span class="p">[</span><span class="n">x</span><span class="p">]</span><span class="o">-</span><span class="n">ca_atom_list</span><span class="p">[</span><span class="n">y</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Total number of Ca-Ca distances&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ca_ca_dist</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Total number of atoms 1825
Number of Ca atoms 203
Total number of Ca-Ca distances 20503
</pre></div>
</div>
</div>
</div>
<p>Next we need to convert the list of Ca-Ca distance to a matrix of size 203x203. The <code class="docutils literal notranslate"><span class="pre">squareform()</span></code> function in the scipy library can be used for this. Once the matix is prepared, we can plot the heatmap for the distances. Also, the contact map can be generated by appling a distance cutoff.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">squareform</span>
<span class="n">dist_mat</span> <span class="o">=</span> <span class="n">squareform</span><span class="p">(</span><span class="n">ca_ca_dist</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dist_mat</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(203, 203)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">dist_mat</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;RdBu_r&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;Distance ($\AA$)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/BIO_PDB_34_0.png" src="_images/BIO_PDB_34_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cont_map</span> <span class="o">=</span> <span class="n">dist_mat</span>
<span class="n">cont_map</span><span class="p">[</span><span class="n">cont_map</span><span class="o">&gt;</span><span class="mi">7</span><span class="p">]</span><span class="o">=</span><span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">cont_map</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;YlGn&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">colorbar</span><span class="p">(</span><span class="n">label</span><span class="o">=</span><span class="s2">&quot;Distance ($\AA$)&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/BIO_PDB_36_0.png" src="_images/BIO_PDB_36_0.png" />
</div>
</div>
<p>To find contact residues based on distance cutoff, <code class="docutils literal notranslate"><span class="pre">NeighborSearch()</span></code> can be used. We frst need to create an atom selection that has the desired residues for which contacting residues are required. E.g., to create an atom selection having all atoms in a chain, we can use the <code class="docutils literal notranslate"><span class="pre">unfold_entities()</span></code> as below. Next we’ll create a <code class="docutils literal notranslate"><span class="pre">NeighborSearch</span></code> object having all the atoms in the other chain. Then we’ll find the residues contacting both the chains.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">A_atoms</span> <span class="o">=</span> <span class="n">Selection</span><span class="o">.</span><span class="n">unfold_entities</span><span class="p">(</span><span class="n">structure</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;A&quot;</span><span class="p">],</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span> <span class="c1">#entity list, target level (A,R,C,M,S)</span>
<span class="n">B_atoms</span> <span class="o">=</span> <span class="n">Selection</span><span class="o">.</span><span class="n">unfold_entities</span><span class="p">(</span><span class="n">structure</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;B&quot;</span><span class="p">],</span> <span class="s2">&quot;A&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ns</span> <span class="o">=</span> <span class="n">NeighborSearch</span><span class="p">(</span><span class="n">A_atoms</span><span class="p">)</span>

<span class="n">dist_cutoff</span><span class="o">=</span><span class="mi">3</span>
<span class="n">contacts_B</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">B_atoms</span><span class="p">:</span>
    <span class="n">close_atoms</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">close_atoms</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">close_atoms</span><span class="p">:</span>
            <span class="n">contacts_B</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get unique elements in contacts_B</span>
<span class="nb">set</span><span class="p">(</span><span class="n">contacts_B</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{20,
 24,
 51,
 53,
 115,
 118,
 119,
 120,
 201,
 307,
 310,
 314,
 339,
 341,
 349,
 358,
 387,
 403}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ns</span> <span class="o">=</span> <span class="n">NeighborSearch</span><span class="p">(</span><span class="n">B_atoms</span><span class="p">)</span>

<span class="n">contacts_A</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">atom</span> <span class="ow">in</span> <span class="n">A_atoms</span><span class="p">:</span>
    <span class="n">close_atoms</span> <span class="o">=</span> <span class="n">ns</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="n">atom</span><span class="o">.</span><span class="n">coord</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">close_atoms</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">a</span> <span class="ow">in</span> <span class="n">close_atoms</span><span class="p">:</span>
            <span class="n">contacts_A</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">get_parent</span><span class="p">()</span><span class="o">.</span><span class="n">get_id</span><span class="p">()[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#get unique elements in contacts_B</span>
<span class="nb">set</span><span class="p">(</span><span class="n">contacts_A</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{18,
 20,
 50,
 51,
 53,
 115,
 118,
 119,
 120,
 213,
 214,
 262,
 272,
 275,
 282,
 285,
 287,
 290}
</pre></div>
</div>
</div>
</div>
<p>Visualizing structures using nglview library. To install this library run <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">nglview</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">nglview</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "2a0dac0e36a2445989b2e306a83f73e0", "version_major": 2, "version_minor": 0}
</script></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">view</span> <span class="o">=</span> <span class="n">nglview</span><span class="o">.</span><span class="n">show_structure_file</span><span class="p">(</span><span class="s2">&quot;7jtl.pdb&quot;</span><span class="p">)</span>
<span class="n">view</span><span class="o">.</span><span class="n">add_representation</span><span class="p">(</span><span class="s1">&#39;cartoon&#39;</span><span class="p">,</span> <span class="n">selection</span><span class="o">=</span><span class="s1">&#39;protein&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;blue&#39;</span><span class="p">)</span>
<span class="n">view</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id": "36bc62dc6bdb4d1f944590b960b7bc45", "version_major": 2, "version_minor": 0}
</script></div>
</div>
</div>
</div>
<div class="section" id="binana">
<h1>Binana<a class="headerlink" href="#binana" title="Permalink to this headline">#</a></h1>
<p>BINANA (BINding ANAlyzer) is a python library that can be used for protein-ligand binding analysis. The source code is available for <a class="reference external" href="https://git.durrantlab.pitt.edu/jdurrant/binana">download</a>. After download the file, unzip the archive and add the path of the location where binana is there to the sys.path as shown below. You should now be able to import binana.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;&lt;full_path&gt;/binana-2.1/python/&quot;</span><span class="p">)</span><span class="c1">#replace &lt;full_path&gt; with actual path.</span>
<span class="kn">import</span> <span class="nn">binana</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># pdbqt files are available in examples dir within binana</span>
<span class="n">ligand</span><span class="p">,</span> <span class="n">receptor</span> <span class="o">=</span> <span class="n">binana</span><span class="o">.</span><span class="n">load_ligand_receptor</span><span class="o">.</span><span class="n">from_files</span><span class="p">(</span><span class="s2">&quot;ligand.pdbqt&quot;</span><span class="p">,</span> \
                                                          <span class="s2">&quot;receptor.pdbqt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The interactions class has a set of functions for retriving different types of inter-molecular interactions. E.g., getting a list of hydrogen bonds between the receptor and the ligand using <code class="docutils literal notranslate"><span class="pre">get_hydrogen_bonds()</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">hbond_inf</span> <span class="o">=</span> <span class="n">binana</span><span class="o">.</span><span class="n">interactions</span><span class="o">.</span><span class="n">get_hydrogen_bonds</span><span class="p">(</span><span class="n">ligand</span><span class="p">,</span> <span class="n">receptor</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">hbond_label</span> <span class="ow">in</span> <span class="n">hbond_inf</span><span class="p">[</span><span class="s2">&quot;labels&quot;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hbond_label</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(&#39;A:CHT(1):N1(14)&#39;, &#39;A:CHT(1):H1(16)&#39;, &#39;A:ASP(157):OD2(285)&#39;, &#39;LIGAND&#39;, {&#39;distance&#39;: 2.6500811308335455, &#39;angle&#39;: 16.087842801376098})
(&#39;A:CHT(1):O6(22)&#39;, &#39;A:ASN(156):2HD2(276)&#39;, &#39;A:ASN(156):ND2(274)&#39;, &#39;RECEPTOR&#39;, {&#39;distance&#39;: 2.9006795755477723, &#39;angle&#39;: 35.51562311681741})
</pre></div>
</div>
</div>
</div>
<p>We can also use <code class="docutils literal notranslate"><span class="pre">get_all_interactions()</span></code> to get a dictionary of all the different types of inter-molecular interactions.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_inf</span> <span class="o">=</span> <span class="n">binana</span><span class="o">.</span><span class="n">interactions</span><span class="o">.</span><span class="n">get_all_interactions</span><span class="p">(</span><span class="n">ligand</span><span class="p">,</span> <span class="n">receptor</span><span class="p">)</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_inf</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>dict_keys([&#39;closest&#39;, &#39;close&#39;, &#39;electrostatic_energies&#39;, &#39;active_site_flexibility&#39;, &#39;hydrophobics&#39;, &#39;hydrogen_bonds&#39;, &#39;halogen_bonds&#39;, &#39;ligand_atom_types&#39;, &#39;pi_pi&#39;, &#39;cat_pi&#39;, &#39;salt_bridges&#39;, &#39;metal_coordinations&#39;, &#39;ligand_rotatable_bonds&#39;])
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_inf</span><span class="p">[</span><span class="s2">&quot;hydrogen_bonds&quot;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>{&#39;counts&#39;: {&#39;HDONOR_LIGAND_SIDECHAIN_OTHER&#39;: 1,
  &#39;HDONOR_RECEPTOR_SIDECHAIN_OTHER&#39;: 1},
 &#39;mol&#39;: &lt;binana._structure.mol.Mol at 0x1232c4be0&gt;,
 &#39;labels&#39;: [(&#39;A:CHT(1):N1(14)&#39;,
   &#39;A:CHT(1):H1(16)&#39;,
   &#39;A:ASP(157):OD2(285)&#39;,
   &#39;LIGAND&#39;,
   {&#39;distance&#39;: 2.6500811308335455, &#39;angle&#39;: 16.087842801376098}),
  (&#39;A:CHT(1):O6(22)&#39;,
   &#39;A:ASN(156):2HD2(276)&#39;,
   &#39;A:ASN(156):ND2(274)&#39;,
   &#39;RECEPTOR&#39;,
   {&#39;distance&#39;: 2.9006795755477723, &#39;angle&#39;: 35.51562311681741})]}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_data</span> <span class="o">=</span> <span class="n">binana</span><span class="o">.</span><span class="n">output</span><span class="o">.</span><span class="n">dictionary</span><span class="o">.</span><span class="n">collect_all</span><span class="p">(</span><span class="n">all_inf</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">all_data</span><span class="p">[</span><span class="s2">&quot;hydrogenBonds&quot;</span><span class="p">]))</span>
<span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_data</span><span class="p">[</span><span class="s2">&quot;hydrogenBonds&quot;</span><span class="p">]:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>2
{&#39;ligandAtoms&#39;: [{&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 1, &#39;resName&#39;: &#39;CHT&#39;, &#39;atomName&#39;: &#39;N1&#39;, &#39;atomIndex&#39;: 14}, {&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 1, &#39;resName&#39;: &#39;CHT&#39;, &#39;atomName&#39;: &#39;H1&#39;, &#39;atomIndex&#39;: 16}], &#39;receptorAtoms&#39;: [{&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 157, &#39;resName&#39;: &#39;ASP&#39;, &#39;atomName&#39;: &#39;OD2&#39;, &#39;atomIndex&#39;: 285}], &#39;metrics&#39;: {&#39;distance&#39;: 2.6500811308335455, &#39;angle&#39;: 16.087842801376098}}
{&#39;ligandAtoms&#39;: [{&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 1, &#39;resName&#39;: &#39;CHT&#39;, &#39;atomName&#39;: &#39;O6&#39;, &#39;atomIndex&#39;: 22}], &#39;receptorAtoms&#39;: [{&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 156, &#39;resName&#39;: &#39;ASN&#39;, &#39;atomName&#39;: &#39;ND2&#39;, &#39;atomIndex&#39;: 274}, {&#39;chain&#39;: &#39;A&#39;, &#39;resID&#39;: 156, &#39;resName&#39;: &#39;ASN&#39;, &#39;atomName&#39;: &#39;2HD2&#39;, &#39;atomIndex&#39;: 276}], &#39;metrics&#39;: {&#39;distance&#39;: 2.9006795755477723, &#39;angle&#39;: 35.51562311681741}}
</pre></div>
</div>
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="Phylo.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Phylogenetic analysis</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Manish Datt (manish@bioinfo.guru)<br/>
  
      &copy; Copyright 2022 bioinfo.guru.<br/>
    <div class="extra_footer">
      <a href="https://bioinfo.guru">Online course catalogue</a>
    </div>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>